### PR TITLE
Add Delayed functionality to shinq

### PR DESCRIPTION
**Pull request by aeroastro**
@ryopeko 

# Summary

I would like to implement ActiveJob Delayed functionality through Q4M conditional subscription.

This PR contains commit which is to be merged by #14. If you can merge the PR, I can clean the changed files.

# Background

Sometimes we need Delayed functionality for ActiveJob in cases like

* Scheduling requirement like scheduled Push Notification
* Exponential backoff for unexpected system failure
* API limit on external services

Some ActiveJob adapters have implemented this and others have not.

Fortunately, Q4M has Conditional Subscription which is suited for Delayed or scheduled job.

# Implementation

* Add `scheduled_at` for scheduled date. (0 for default immediate job)
  * We use Unix time instead of datetime for a known issue on Q4M
* `ShinqAdapter#enqueue_at` for ActiveJob::QueueAdapter interface
* Maintain backward compatibility by fetching database column names
  * Alter table on Q4M could easily corrupt data with OWNER-mode connection.
  * Fetch database column before fork for performance issue.

# References
QueueAdapters
http://api.rubyonrails.org/classes/ActiveJob/QueueAdapters.html

implementation of Delayed Job by sidekiq
https://github.com/rails/rails/blob/4-2-stable/activejob/lib/active_job/queue_adapters/sidekiq_adapter.rb#L28-L35

Conditional Subscription
https://q4m.github.io/tutorial.html

conditional subscription fails against a column after than a DATETIME column (MySQL 5.6)
https://github.com/q4m/q4m/issues/9

ServerEngine Worker module interface
https://github.com/treasure-data/serverengine#worker-module

# Future work

After this PR is merged, I am going to implement a functionality to share the same queue between multiple Jobs. This can dramatically improves CPU and memory efficiency when there are many kind of workers and they are unevenly loaded.

### CODE REVIEW

**Comment by aeroastro** [comment]
It seems like we need to stop some `before` procedures when testing in Travis.

**Comment by aeroastro** [comment]
Travis MySQL problems were fixed by using InnoDB engine on Travis and QUEUE engine on the other environments.
All the test outside the integration_spec do not depend on QUEUE engine.

**Comment by aeroastro** [comment]
I've resolved conflict by rebasing.
Also, I've found a tiny bug on `scheduled_at` and fixed it. (Make queue run at scheduled_at not after scheduled_at)

File: spec/shinq_spec.rb [hunk 1]
```diff
@@ -9,6 +9,10 @@ def shinq_class
 
 describe Shinq do
   subject { shinq_class }
+  after do
+    Shinq.clear_all_connections!
+    shinq_class.clear_all_connections!
```
Comment (ryopeko) [inline spec/shinq_spec.rb]: looks same code

**Comment by ryopeko** [comment]
Please check my comment.
Looks good to me.

Thank you for added to some tests.
Great work!!!!

File: spec/shinq_spec.rb (hunk 1 above)
Comment (aeroastro) [inline spec/shinq_spec.rb]: These two lines seem the same codes but actually are not.
Former one clears connections stored in `Shinq`, while latter one clears connections stored in duped `Shinq`.

The reason why the former one was required in addition to the latter one is that `shinq_class` was called many times and those duped classes shares the same object `@connections = {}` which were defined in `Shinq` in prior test. Therefore, `Shinq.clear_all_connections!` cleared not only `Shinq` connections but also all the connections defined in `Shinq.dup`.

I fixed duping strategy on rspec to resolve this problem. Now the duped class is consistent within each example with all the instance variables initialized (removed).

**Comment by aeroastro** [comment]
Thank you for reviewing the code 🐱 
I've fixed the problem in rspec.

File: spec/shinq_spec.rb (hunk 1 above)
Comment (ryopeko) [reply spec/shinq_spec.rb]: OK. thanks.
Maybe we use `described_class` .

**Comment by ryopeko** [comment]
I will release as v1.0.0.rc1.
Please use this version.
Version bump as v1.0.0 if no problem in this version.

**Comment by ryopeko** [comment]
@aeroastro 
I've just released it.
https://rubygems.org/gems/shinq/versions/1.0.0.rc1

**Comment by aeroastro** [comment]
Thank you 🐱

**Pull request merged**